src/arch/io/network.cc

```diff
@@ -621,6 +621,430 @@ void linux_tcp_conn_t::on_event(int /* events */) {
     event_watcher->stop_watching_for_errors();
 }
 
+/* tls_conn_wrapper_t wraps a TLS connection and safely releases it in
+exceptional cases if it is not reclaimed before execution leaves its scope. */
+class tls_conn_wrapper_t {
+public:
+    tls_conn_wrapper_t(SSL *_conn) {
+        conn = _conn;
+    }
+
+    ~tls_conn_wrapper_t() {
+        SSL_free(conn);
+    }
+
+    // Transfer ownership of the connection to the caller.
+    void reclaim() {
+        conn = NULL;
+    }
+
+private:
+    SSL *conn;
+};
+
+/* This is the client version of the constructor. The base class constructor
+will establish a TCP connection to the peer at the given host:port and then we
+wrap the tcp connection in TLS using the configuration in the given tls_ctx. */
+linux_secure_tcp_conn_t::linux_secure_tcp_conn_t(SSL_CTX *tls_ctx, const ip_address_t &host, int port, signal_t *interruptor, int local_port) THROWS_ONLY(connect_failed_exc_t, interrupted_exc_t) :
```

Comment:
Another small style thing that you couldn't possibly have known because our style guide wasn't available anywhere: This line should be wrapped to an 89 character line length (maximum).